### ISSUE TITLE
Automatic Node/Disk Setup

### ISSUE THREAD

**Issue by kevinsingapore** [longhorn/longhorn#1053]
how to define different default data path in different nodes when initiating longhorn service?

**Comment by kevinsingapore** [comment]
> how to define different default data path in different nodes when initiating longhorn service?

by using the default longhorn.yaml.

**Comment by yasker** [comment]
Please check https://github.com/longhorn/longhorn/blob/master/docs/customized-default-setting.md

**Comment by kevinsingapore** [comment]
> Please check https://github.com/longhorn/longhorn/blob/master/docs/customized-default-setting.md

thank u!
the url you posted is not my wanted. it's only the same default storage path on all nodes;
on my way, it means to define different file directory path on different nodes as the longhorn storage path，sush as : 
(1)node1: /var/lib/rancher/longhorn-example；
(2)node2: /home/rancher/longhorn-example;
(3)...

**Comment by yasker** [comment]
I see. Currently this cannot be done through default setup. You need to deploy Longhorn then update the disk to get that. 

I am thinking probably we can use labels/annotations on the Kubernetes nodes for that? Similiar to what we're planning with https://github.com/longhorn/longhorn/issues/991

**Comment by yasker** [comment]
Design

1. Kubernetes doesn't have the concept of `disk`, so we need to use the Kubernetes node object instead.
2. We can support annotations to the node object.
e.g.
``` 
`disk.longhorn.io/creation-config: {"disks": {{"path": "/mnt/disk1", "storageReserved":"1G"}, {"path": "/mnt/disk2, "tag": {list of tags}}}`
`node.longhorn.io/creation-config: {tags: {"xxx", }}`
```
3. Longhorn will not remove the disk if the disk is removed from configuration, or modify the node annotation in any way.
4. The tag specified in the node will be added as additional tags if they don't exist yet
5. Longhorn will use the bigger number of storageReserved vs the current value of storageReserved.
6. This feature doesn't intervene with default disk creation feature

**Comment by meldafrawi** [comment]
Validation: **Partially FAILED**

Steps to test:
- Create 3 node k8s cluster
- Create `/home/longhorn` directory on all 3 nodes
- add new disk to each node, format it with `ext4`, and mount it to `/mnt/disk`
- Use the following label and annotations for nodes

node01 & node03
```
labels:
    node.longhorn.io/create-default-disk: "config"
annotations:
    node.longhorn.io/default-disks-config:
        '[{"path":"/home/longhorn","allowScheduling":true, "tags":["ssd", "fast"]},
          {"path":"/mnt/disk","allowScheduling":true,"storageReserved":1024,"tags":["ssd","fast"]}]'
    node.longhorn.io/default-node-tags: '["fast", "storage"]'
```

node 02
```
labels:
    node.longhorn.io/create-default-disk: "config"
annotations:
    node.longhorn.io/default-disks-config:
        '[{"path":"/home/longhorn","allowScheduling":true, "tags":["hdd", "slow"]},
          {"path":"/mnt/disk","allowScheduling":true,"storageReserved":1024,"tags":["hdd","slow"]}]'
    node.longhorn.io/default-node-tags: '["slow","storage"]'
```

- Set  `create-default-disk-labeled-nodes: True` in `longhorn-default-setting` config map
- Deploy Longhorn (master)

**Expected result**: in UI `Node` page,  for each node disks and tags should match the specified annotations annotations (FAILED)

Actual Result: 
- In Settings page, `Create Default Disk on Labeled Nodes` should be enabled. (FAILED).
- Node tags match tags specified in annotation  (PASSED).
- default disk with `/var/lib/longhorn/` path is created, no tags is specified (FAILED).

**Comment by shuo-wu** [comment]
The failure is actually caused by the issue #1133.

Please set `create-default-disk-labeled-nodes` to `true` for this test.

**Comment by meldafrawi** [comment]
Validation: **PASSED**

fixed by PR in issue #1133.
